### ISSUE TITLE
Existing URL-encoded query params are encoded again when URL contains a query string

### ISSUE THREAD

**Issue by sjwj** [OpenFeign/feign#662]
It appears as though feign is re-URL-encoding all existing parameters (e.g. those provided through RequestTemplate.query()) when inserting a URL that contains a query string.

I've written the following tests that demonstrate the problem:

```java package feign;

import org.junit.Test;

import static feign.assertj.FeignAssertions.assertThat;
import static java.util.Arrays.asList;
import static org.assertj.core.data.MapEntry.entry;

public class RequestTemplateEncodingTest {

    @Test
    // fails
    public void insertWithQueryStringDoesNotReencodeExistingParams() throws Exception {
        RequestTemplate template = new RequestTemplate().method("GET")
            .append("/domains/1001/records")
            .query("name", "de,nominator.io")
            .query("type", "CNAME");

        template.insert(0, "https://host/v1.0/1234?provider=foo%20bar");

        assertThat(template)
            .hasUrl("https://host/v1.0/1234/domains/1001/records")
            .hasQueries(
                entry("provider", asList("foo bar")),
                entry("name", asList("de,nominator.io")),
                entry("type", asList("CNAME"))
            );
    }

    @Test
    // passes
    public void insertWithoutQueryStringDoesNotReencodeExistingParams() throws Exception {
        RequestTemplate template = new RequestTemplate().method("GET")
            .append("/domains/1001/records")
            .query("name", "de,nominator.io")
            .query("type", "CNAME");

        template.insert(0, "https://host/v1.0/1234");

        assertThat(template)
            .hasUrl("https://host/v1.0/1234/domains/1001/records")
            .hasQueries(
                entry("name", asList("de,nominator.io")),
                entry("type", asList("CNAME"))
            );
    }

}
```

I had a bit of a look through the code and the problem seems to stem from:
https://github.com/OpenFeign/feign/blob/7bea518/core/src/main/java/feign/RequestTemplate.java#L557

```java
  /**
   * if there are any query params in the URL, this will extract them out.
   */
  private StringBuilder pullAnyQueriesOutOfUrl(StringBuilder url) {
    // parse out queries
    int queryIndex = url.indexOf("?");
    if (queryIndex != -1) {
      String queryLine = url.substring(queryIndex + 1);
      Map<String, Collection<String>> firstQueries = parseAndDecodeQueries(queryLine);
      if (!queries.isEmpty()) {
        firstQueries.putAll(queries);
        queries.clear();
      }
      //Since we decode all queries, we want to use the
      //query()-method to re-add them to ensure that all
      //logic (such as url-encoding) are executed, giving
      //a valid queryLine()
      for (String key : firstQueries.keySet()) {
        Collection<String> values = firstQueries.get(key);
        if (allValuesAreNull(values)) {
          //Queries where all values are null will
          //be ignored by the query(key, value)-method
          //So we manually avoid this case here, to ensure that
          //we still fulfill the contract (ex. parameters without values)
          queries.put(urlEncode(key), values);
        } else {
          query(key, values);
        }

      }
      return new StringBuilder(url.substring(0, queryIndex));
    }
    return url;
  }
```

Feign is extracting (and decoding) all the query parameters out of the provided URL into a Map, [here](https://github.com/OpenFeign/feign/blob/7bea518/core/src/main/java/feign/RequestTemplate.java#L562) then subsequently adding all of the existing (already encoded) parameters into the same Map [here](https://github.com/OpenFeign/feign/blob/7bea518/core/src/main/java/feign/RequestTemplate.java#L564)

All of the values in the map are then re-URL encoded (transitively through the query() method) [here](https://github.com/OpenFeign/feign/blob/7bea518/core/src/main/java/feign/RequestTemplate.java#L580)

I believe the desired behavior is that feign should extract the parameters out of the provided URL and merge them with the existing set of parameters -- having the existing parameters maintain their current values.

(This is likely the same issue as https://github.com/OpenFeign/feign/issues/483).

**Comment by kdavisk6** [comment]
See #778

**Comment by kdavisk6** [comment]
Closed via #778
